README.md
# Awesome Number Picker

## Usage

To use plugin, just import package `import 'package:awesome_number_picker/awesome_number_picker.dart';`
and use IntegerNumberPicker() or DecimalNumberPicker()

## Features

- Pure dart code.
- Simple and easy to use.
- Customizable design.

## Demo
![image](https://user-images.githubusercontent.com/22174249/164450869-8c3b641a-5811-4d52-9463-6c77f93a3016.png)

```

import 'package:awesome_number_picker/awesome_number_picker.dart';
import 'package:flutter/material.dart';

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  int integerValue = 0;
  double decimalValue = 0;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Awesome Number Picker")),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const Text("Integer Value"),
              Text(integerValue.toString()),
            ],
          ),
          SizedBox(
            height: 150,
            child: IntegerNumberPicker(
              initialValue: 5,
              minValue: 0,
              maxValue: 10,
              onChanged: (i) => setState(() {
                integerValue = i;
              }),
            ),
          ),
          SizedBox(
            height: 150,
            child: IntegerNumberPicker(
              initialValue: 75,
              minValue: 50,
              maxValue: 100,
              axis: Axis.horizontal,
              onChanged: (i) => setState(() {
                integerValue = i;
              }),
            ),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const Text("Decimal Value"),
              Text(decimalValue.toStringAsFixed(3)),
            ],
          ),
          SizedBox(
            height: 150,
            child: DecimalNumberPicker(
              initialValue: 25.75,
              minValue: 0,
              maxValue: 100,
              decimalPrecision: 3,
              otherItemsDecoration: BoxDecoration(
                boxShadow: [
                  BoxShadow(
                    color: Colors.grey.withOpacity(.5),
                    blurRadius: 20.0, // soften the shadow
                    spreadRadius: 0.0, //extend the shadow
                  )
                ],
              ),
              pickedItemDecoration: BoxDecoration(
                boxShadow: [
                  BoxShadow(
                    color: Colors.blue.withOpacity(.5),
                    blurRadius: 20.0, // soften the shadow
                    spreadRadius: 0.0, //extend the shadow
                  )
                ],
              ),
              onChanged: (i) => setState(() {
                decimalValue = i;
              }),
            ),
          ),
        ],
      ),
    );
  }
}


```

## Additional information


CHANGELOG.md
## 0.0.1

* NEW: Initial Release.

## 0.0.2

* Some fixes

## 0.0.3

* Add DecimalNumberPicker

## 0.0.4

* For some fixes


example/lib/decimal_test.dart
import 'package:awesome_number_picker/awesome_number_picker.dart';
import 'package:flutter/material.dart';

class DecimalTest extends StatefulWidget {
  const DecimalTest({Key? key}) : super(key: key);

  @override
  State<DecimalTest> createState() => _DecimalTestState();
}

class _DecimalTestState extends State<DecimalTest> {
  double value = 0;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Text(value.toString()),
          SizedBox(
            height: 200,
            child: DecimalNumberPicker(
                initialValue: 3.1,
                minValue: 0,
                maxValue: 10,
                decimalPrecision: 2,
                onChanged: (i) {
                  setState(() {
                    value = i;
                  });
                }),
          ),
        ],
      ),
    );
  }
}


example/lib/home_page.dart
import 'package:awesome_number_picker/awesome_number_picker.dart';
import 'package:flutter/material.dart';

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  int integerValue = 0;
  double decimalValue = 0;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Awesome Number Picker")),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const Text("Integer Value"),
              Text(integerValue.toString()),
            ],
          ),
          SizedBox(
            height: 150,
            child: IntegerNumberPicker(
              initialValue: 5,
              minValue: 0,
              maxValue: 10,
              onChanged: (i) => setState(() {
                integerValue = i;
              }),
            ),
          ),
          SizedBox(
            height: 150,
            child: IntegerNumberPicker(
              initialValue: 75,
              minValue: 50,
              maxValue: 100,
              axis: Axis.horizontal,
              onChanged: (i) => setState(() {
                integerValue = i;
              }),
            ),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const Text("Decimal Value"),
              Text(decimalValue.toStringAsFixed(3)),
            ],
          ),
          SizedBox(
            height: 150,
            child: DecimalNumberPicker(
              initialValue: 25.75,
              minValue: 0,
              maxValue: 100,
              decimalPrecision: 3,
              otherItemsDecoration: BoxDecoration(
                boxShadow: [
                  BoxShadow(
                    color: Colors.grey.withOpacity(.5),
                    blurRadius: 20.0, // soften the shadow
                    spreadRadius: 0.0, //extend the shadow
                  )
                ],
              ),
              pickedItemDecoration: BoxDecoration(
                boxShadow: [
                  BoxShadow(
                    color: Colors.blue.withOpacity(.5),
                    blurRadius: 20.0, // soften the shadow
                    spreadRadius: 0.0, //extend the shadow
                  )
                ],
              ),
              onChanged: (i) => setState(() {
                decimalValue = i;
              }),
            ),
          ),
        ],
      ),
    );
  }
}


lib/src/decimal_number_picker.dart
import 'dart:math';

import 'package:awesome_number_picker/awesome_number_picker.dart';
import 'package:flutter/material.dart';

class DecimalNumberPicker extends StatefulWidget {
  /// Selected value for Start position - #initialValue cant be bigger than #maxValue and cant be smaller than #minValue
  final double initialValue;

  /// Min value user can pick - #minValue cant be bigger than #initialValue amd #maxValue
  final int minValue;

  /// Max value user can pick
  final int maxValue;

  ///
  final int decimalPrecision;

  /// if user interaction picker this methods return selected value
  final ValueChanged<double> onChanged;

  /// height and width of single item for vertical.
  final double? integerPartSize;
  final double? decimalPartSize;

  /// Decoration to apply to picked number where the selected value is placed
  final BoxDecoration? pickedItemDecoration;

  /// Decoration to apply non-picked numbers
  final BoxDecoration? otherItemsDecoration;

  /// Style of picked number. If null, it uses Theme's headline5 with primaryColor
  final TextStyle? pickedItemTextStyle;

  /// Style of non-picked numbers. If null, it uses Theme's bodyText2
  final TextStyle? otherItemsTextStyle;

  const DecimalNumberPicker({
    Key? key,
    required this.initialValue,
    required this.minValue,
    required this.maxValue,
    required this.decimalPrecision,
    required this.onChanged,
    this.integerPartSize,
    this.decimalPartSize,
    this.pickedItemDecoration,
    this.otherItemsDecoration,
    this.pickedItemTextStyle,
    this.otherItemsTextStyle,
  })  : assert(initialValue < maxValue),
        assert(initialValue >= minValue),
        assert(minValue < maxValue),
        super(key: key);

  @override
  State<DecimalNumberPicker> createState() => _DecimalNumberPickerState();
}

class _DecimalNumberPickerState extends State<DecimalNumberPicker> {
  late int returnIntegerPart;
  late int returnDecimalPart;
  late int integerPartInit;
  late int decimalPartInit;
  late int returnValueIntegerPart;
  late int returnValueDecimalPart;

  @override
  void initState() {
    super.initState();
    String initialString = (widget.initialValue - widget.minValue)
        .toStringAsFixed(widget.decimalPrecision);
    String integerPartInitStr = initialString.split(".")[0];
    String decimalPartInitStr = initialString.split(".")[1];
    integerPartInit = int.parse(integerPartInitStr);
    decimalPartInit = int.parse(decimalPartInitStr);

    print(integerPartInit);
    print(decimalPartInit);
    returnIntegerPart = integerPartInit;
    returnDecimalPart = decimalPartInit;
  }

  returnValue() {
    String temp = "$returnIntegerPart.$returnDecimalPart";

    widget.onChanged(double.parse(temp));
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          SizedBox(
            width: widget.integerPartSize ?? 40,
            child: IntegerNumberPicker(
                initialValue: integerPartInit,
                minValue: widget.minValue,
                maxValue: widget.maxValue,
                size: widget.integerPartSize,
                pickedItemDecoration: widget.pickedItemDecoration,
                otherItemsDecoration: widget.otherItemsDecoration,
                pickedItemTextStyle: widget.pickedItemTextStyle,
                otherItemsTextStyle: widget.otherItemsTextStyle,
                onChanged: (i) {
                  returnIntegerPart = i;
                  setState(() {
                    returnValue();
                  });
                }),
          ),
          SizedBox.shrink(),
          SizedBox(
            width: widget.decimalPartSize ?? 50,
            child: IntegerNumberPicker(
                initialValue: decimalPartInit,
                minValue: widget.minValue,
                maxValue: pow(10, widget.decimalPrecision).toInt(),
                size: widget.decimalPartSize ?? 50,
                pickedItemDecoration: widget.pickedItemDecoration,
                otherItemsDecoration: widget.otherItemsDecoration,
                pickedItemTextStyle: widget.pickedItemTextStyle,
                otherItemsTextStyle: widget.otherItemsTextStyle,
                onChanged: (i) {
                  returnDecimalPart = i;
                  setState(() {
                    returnValue();
                  });
                }),
          ),
        ],
      ),
    );
  }
}


lib/src/integer_number_picker.dart
import 'dart:math';

import 'package:flutter/material.dart';

class IntegerNumberPicker extends StatefulWidget {
  /// Selected value for Start position - #initialValue cant be bigger than #maxValue and cant be smaller than #minValue
  final int initialValue;

  /// Min value user can pick - #minValue cant be bigger than #initialValue amd #maxValue
  final int minValue;

  /// Max value user can pick
  final int maxValue;

  /// if user interaction picker this methods return selected value
  final ValueChanged<int> onChanged;

  /// Direction of scrolling.
  final Axis? axis;

  /// height and width of single item for vertical.
  final double? size;

  /// Decoration to apply to picked number where the selected value is placed
  final BoxDecoration? pickedItemDecoration;

  /// Decoration to apply non-picked numbers
  final BoxDecoration? otherItemsDecoration;

  /// Style of picked number. If null, it uses Theme's headline5 with primaryColor
  final TextStyle? pickedItemTextStyle;

  /// Style of non-picked numbers. If null, it uses Theme's bodyText2
  final TextStyle? otherItemsTextStyle;

  const IntegerNumberPicker({
    Key? key,
    required this.initialValue,
    required this.minValue,
    required this.maxValue,
    required this.onChanged,
    this.axis,
    this.size,
    this.pickedItemDecoration,
    this.otherItemsDecoration,
    this.pickedItemTextStyle,
    this.otherItemsTextStyle,
  })  : assert(initialValue < maxValue),
        assert(initialValue >= minValue),
        assert(minValue < maxValue),
        super(key: key);

  @override
  State<IntegerNumberPicker> createState() => _IntegerNumberPickerState();
}

class _IntegerNumberPickerState extends State<IntegerNumberPicker> {
  late FixedExtentScrollController _controller;
  late int returnValue;

  @override
  void initState() {
    super.initState();
    returnValue = widget.initialValue - widget.minValue;
    _controller = FixedExtentScrollController(initialItem: returnValue);
  }

  @override
  Widget build(BuildContext context) {
    final pickedItemStyle = BoxDecoration(
        border: Border.all(color: Theme.of(context).primaryColor, width: 3),
        borderRadius: const BorderRadius.all(Radius.circular(360)));

    final otherItemsStyle = BoxDecoration(
        border: Border.all(color: Theme.of(context).unselectedWidgetColor),
        borderRadius: const BorderRadius.all(Radius.circular(360)));

    final pickedItemTextStyle = Theme.of(context)
        .textTheme
        .titleMedium!
        .copyWith(color: Theme.of(context).primaryColor);
    final otherItemsTextStyle = Theme.of(context)
        .textTheme
        .titleSmall!
        .copyWith(color: Theme.of(context).unselectedWidgetColor);

    return Transform.rotate(
      angle: widget.axis == Axis.vertical || widget.axis == null ? 0 : -pi / 2,
      child: ListWheelScrollView.useDelegate(
        controller: _controller,
        onSelectedItemChanged: ((value) {
          widget.onChanged(value + widget.minValue);
          setState(() {
            returnValue = value;
          });
        }),
        physics: const BouncingScrollPhysics(),
        perspective: 0.01,
        itemExtent: widget.size ?? 40,
        childDelegate: ListWheelChildBuilderDelegate(
          childCount: widget.maxValue - widget.minValue,
          builder: (c, i) => i == returnValue
              ? Container(
                  width: widget.size ?? 40,
                  decoration: widget.pickedItemDecoration ?? pickedItemStyle,
                  child: Transform.rotate(
                    angle: widget.axis == Axis.vertical || widget.axis == null
                        ? 0
                        : pi / 2,
                    child: Center(
                      child: Text((i + widget.minValue).toString(),
                          style: widget.pickedItemTextStyle ??
                              pickedItemTextStyle),
                    ),
                  ),
                )
              : Container(
                  width: widget.size ?? 40,
                  decoration: widget.otherItemsDecoration ?? otherItemsStyle,
                  child: Transform.rotate(
                    angle: widget.axis == Axis.vertical || widget.axis == null
                        ? 0
                        : pi / 2,
                    child: Center(
                      child: Text(
                        (i + widget.minValue).toString(),
                        style:
                            widget.otherItemsTextStyle ?? otherItemsTextStyle,
                      ),
                    ),
                  ),
                ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:awesome_number_picker/awesome_number_picker.dart';
import 'package:example/scroll_behaviour.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Flutter Demo',
      scrollBehavior: CustomScrollBehavior(), //For web Scrolling
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  int integerValue = 0;
  double decimalValue = 0;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Awesome Number Picker")),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const Text("Integer Value"),
              Text(integerValue.toString()),
            ],
          ),
          SizedBox(
            height: 150,
            child: IntegerNumberPicker(
              initialValue: 5,
              minValue: 0,
              maxValue: 10,
              onChanged: (i) => setState(() {
                integerValue = i;
              }),
            ),
          ),
          SizedBox(
            height: 150,
            child: IntegerNumberPicker(
              initialValue: 75,
              minValue: 50,
              maxValue: 100,
              axis: Axis.horizontal,
              onChanged: (i) => setState(() {
                integerValue = i;
              }),
            ),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const Text("Decimal Value"),
              Text(decimalValue.toStringAsFixed(3)),
            ],
          ),
          SizedBox(
            height: 150,
            child: DecimalNumberPicker(
              initialValue: 25.75,
              minValue: 0,
              maxValue: 100,
              decimalPrecision: 3,
              otherItemsDecoration: BoxDecoration(
                boxShadow: [
                  BoxShadow(
                    color: Colors.grey.withOpacity(.5),
                    blurRadius: 20.0, // soften the shadow
                    spreadRadius: 0.0, //extend the shadow
                  )
                ],
              ),
              pickedItemDecoration: BoxDecoration(
                boxShadow: [
                  BoxShadow(
                    color: Colors.blue.withOpacity(.5),
                    blurRadius: 20.0, // soften the shadow
                    spreadRadius: 0.0, //extend the shadow
                  )
                ],
              ),
              onChanged: (i) => setState(() {
                decimalValue = i;
              }),
            ),
          ),
        ],
      ),
    );
  }
}
